### PR TITLE
feat(crd-generator): Add support for validation rules

### PR DESCRIPTION
**Pull request by baloo42**
## Description

Adds support for Kubernetes validation rules to CRDGenerator:

- Validation rules can be added to attributes, accessor methods or classes
- Validation rules can be added to the CustomResource class itself (=the root)
- Validation rules from inherited classes will be discovered

Refers to #4367

## Type of change

 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [x] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [x] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [x] I Added [CHANGELOG](https://github.com/fabric8io/kubernetes-client/blob/main/CHANGELOG.md) entry regarding this change
 - [x] I have implemented unit tests to cover my changes
 - [x] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/main/doc/CHEATSHEET.md) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [x] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

## TODO's

- [x] Fix JavaDoc
- [x] Add more unit tests

### CODE REVIEW

File: crd-generator/api/src/test/java/io/fabric8/crd/generator/v1/JsonSchemaTest.java [hunk 1]
```diff
@@ -177,6 +178,16 @@ void shouldAugmentPropertiesSchemaFromAnnotations() throws JsonProcessingExcepti
     // check ignored fields
     assertFalse(spec.containsKey("ignoredFoo"));
     assertFalse(spec.containsKey("ignoredBar"));
+
+    final JSONSchemaProps kubernetesValidation = spec.get("kubernetesValidation");
+    final List<ValidationRule> kubernetesValidationRules = kubernetesValidation.getXKubernetesValidations();
+    assertNotNull(kubernetesValidationRules);
+    assertEquals(1, kubernetesValidationRules.size());
```
Comment (andreaTP) [inline crd-generator/api/src/test/java/io/fabric8/crd/generator/v1/JsonSchemaTest.java:185]: Do you mind testing the content?

**Review by andreaTP — CHANGES_REQUESTED** [review]
Thanks a lot for this improvement, looks good to me, but I should ask to test the content of the newly generated fields 🙏

File: crd-generator/api/src/test/java/io/fabric8/crd/generator/v1/JsonSchemaTest.java (hunk 1 above)
Comment (baloo42) [reply crd-generator/api/src/test/java/io/fabric8/crd/generator/v1/JsonSchemaTest.java:185]: Yes, I'm not finished with this PR. 

I want to add some more tests which ensures that all parameters are passed through. 
But on my opinion we don't need to test the value content further. It's up to a developer to choose the right validation rules and the CRDGenerator should not contain additional logic to understand those rules.

Comment (andreaTP) [reply crd-generator/api/src/test/java/io/fabric8/crd/generator/v1/JsonSchemaTest.java]: good for me, thanks :+1:

**Review by andreaTP — APPROVED** [review]
LGTM, @baloo42 are you planning to iterate more on this?
I think that's good to go 😄 feel free to remove it from "Draft"

**Comment by baloo42** [comment]
@andreaTP Now I'm finished. I added support for some advanced use cases. 
Please have a look at the added [example](https://github.com/fabric8io/kubernetes-client/tree/c2366a1ef8666a63f2f41f0ab61b0ac2c2b96760/crd-generator/api/src/test/java/io/fabric8/crd/example/k8svalidation) and the expected [result](https://github.com/fabric8io/kubernetes-client/blob/c2366a1ef8666a63f2f41f0ab61b0ac2c2b96760/crd-generator/api/src/test/resources/k8svalidations.samples.fabric8.io-v1.yml).

Now the only feature which is still missing is support for meta-annotations. But that's missing for the other annotations, too. Have you discussed the support of meta-annotations in the CRDGenerator in the past?

File: crd-generator/api/src/test/java/io/fabric8/crd/generator/CRDGeneratorTest.java [hunk 2]
```diff
@@ -545,6 +546,38 @@ void checkGenerationMultipleVersionsOfCRDsIsDeterministic() throws Exception {
     assertTrue(outputDir.delete());
   }
 
+  @Test
+  void checkK8sValidationRules() throws Exception {
+    // generated CRD
+    final File outputDir = Files.createTempDirectory("crd-").toFile();
```
Comment (andreaTP) [inline crd-generator/api/src/test/java/io/fabric8/crd/generator/CRDGeneratorTest.java:552]: If I understand this correctly, moving this test to `crd-generator/test` should remove the need for the "infrastructure" of creating the temp dir etc., can you verify?

File: crd-generator/api/src/test/resources/k8svalidations.samples.fabric8.io-v1.yml [hunk 3]
```diff
@@ -0,0 +1,161 @@
+#
```
Comment (andreaTP) [inline crd-generator/api/src/test/resources/k8svalidations.samples.fabric8.io-v1.yml:1]: is this file used?

File: crd-generator/api/src/test/resources/k8svalidations.samples.fabric8.io-v1beta1.yml [hunk 4]
```diff
@@ -0,0 +1,161 @@
+#
```
Comment (andreaTP) [inline crd-generator/api/src/test/resources/k8svalidations.samples.fabric8.io-v1beta1.yml:1]: is this file used? I fail to find where ...

**Review by andreaTP — CHANGES_REQUESTED** [review]
A couple of minor comments that should slightly simplify the PR, but looks very promising!

**Comment by andreaTP** [comment]
> Now the only feature which is still missing is support for meta-annotations. But that's missing for the other annotations, too. Have you discussed the support of meta-annotations in the CRDGenerator in the past?

not sure I'm getting this one, maybe this is related: #3768

File: crd-generator/api/src/test/java/io/fabric8/crd/generator/CRDGeneratorTest.java (hunk 2 above)
Comment (baloo42) [reply crd-generator/api/src/test/java/io/fabric8/crd/generator/CRDGeneratorTest.java:552]: Yes, it's the same test strategy we already used a few times. For example:

- https://github.com/fabric8io/kubernetes-client/blob/main/crd-generator/api/src/test/java/io/fabric8/crd/generator/CRDGeneratorTest.java#L483
- https://github.com/fabric8io/kubernetes-client/blob/main/crd-generator/api/src/test/java/io/fabric8/crd/generator/CRDGeneratorTest.java#L513

Those kind of tests generate real YAML manifests in the temp directory and compare the resulting files with expected manifests from the test resources.

An already implemented utility assertion method is used to for the comparison, which ignores comments and empty lines:

https://github.com/fabric8io/kubernetes-client/blob/main/crd-generator/api/src/test/java/io/fabric8/crd/generator/CRDGeneratorTest.java#L548

The temp directory will be cleaned up if the test is successfully. If not, the temp directory won't be deleted which makes it easier to debug.

In my opinion those tests are more end-to-end and allow quickly adding edge cases. 
In the future those tests should be refactored to make it more reusable.

File: crd-generator/api/src/test/resources/k8svalidations.samples.fabric8.io-v1.yml (hunk 3 above)
Comment (baloo42) [reply crd-generator/api/src/test/resources/k8svalidations.samples.fabric8.io-v1.yml:1]: Yes, see https://github.com/fabric8io/kubernetes-client/pull/5788#discussion_r1530262059

**Comment by baloo42** [comment]
> > Now the only feature which is still missing is support for meta-annotations. But that's missing for the other annotations, too. Have you discussed the support of meta-annotations in the CRDGenerator in the past?
> 
> not sure I'm getting this one, maybe this is related: #3768

 #3768 is something different. (But should be implemented too in my opinion)

With "meta-annotations" support I mean that a user is able to combine our existing annotations to new annotations like described in the following article with JUnit annotations: 
https://dzone.com/articles/what-are-meta-annotations-in-java

Spring, for example has also support for Meta-Annotations/Composed Annotations:
https://docs.spring.io/spring-framework/reference/core/beans/classpath-scanning.html#beans-meta-annotations

Is this something we want to support in the future? (Is there an easy and performant way to implement it with sundrio?)

File: crd-generator/api/src/test/java/io/fabric8/crd/generator/CRDGeneratorTest.java (hunk 2 above)
Comment (andreaTP) [reply crd-generator/api/src/test/java/io/fabric8/crd/generator/CRDGeneratorTest.java:552]: it's fine, I didn't recall this pattern was already used here, disregard my comment 👍

**Comment by andreaTP** [comment]
> With "meta-annotations" support I mean that a user is able to combine our existing annotations to new annotations

That's interesting, and it has not been discussed previously.
I encourage you to open a separate issue, and I would love to see a strawman on how we can leverage this feature.
If I understand the problem space correctly with meta-annotations we should be able to combine, for example, `io.fabric8.generator.annotation.Max` with `org.hibernate.validator.Max` and that would be a great use-case (also for the `java-generator`)

**Review by andreaTP — APPROVED** [review]
This PR LGTM, thanks again for all the help @baloo42 ! Appreciated 🙏

**Comment by baloo42** [comment]
> > With "meta-annotations" support I mean that a user is able to combine our existing annotations to new annotations
> 
> That's interesting, and it has not been discussed previously. I encourage you to open a separate issue, and I would love to see a strawman on how we can leverage this feature. If I understand the problem space correctly with meta-annotations we should be able to combine, for example, `io.fabric8.generator.annotation.Max` with `org.hibernate.validator.Max` and that would be a great use-case (also for the `java-generator`)

The issue has been created: https://github.com/fabric8io/kubernetes-client/issues/5823
Let's discuss some use cases in this issue. 

(Some use-cases can be useful and can be implemented, some won't work)

**Review by rohanKanojia — APPROVED** [review]

**Review by manusa — APPROVED** [review]
LGTM, thx!
